ffi_derive/src/lib.rs

```diff
@@ -132,6 +133,73 @@
+//! crate. We support that, but because the type information that backs the alias isn't available at
+//! the time procedural macros run, we require some additional configuration in both the module that
+//! defines the alias, and on the type whose fields reference the alias.
+//!
+//! ### Remote alias definitions
+//!
+//! When a module defines aliases that may be referenced by a type that derives an FFI, the
+//! `alias_resolution` attribute macro needs to be run on it in order to populate the definitions of
+//! those aliases somewhere so that we can reference them when resolving the underlying types of
+//! alias references. The macro invocation also needs to define a unique string for the module
+//! (which we refer to internally as the `resolution_key`). This will be used with a helper
+//! attribute on types that derive an FFI so that we can identify the source where their aliases are
+//! defined.
+//!
+//! Invoking the alias resolution macro on a module looks like this:
+//! ```
+//! #[ffi_derive::alias_resolution(some_unique_string)]
+//! mod aliases_here {
+//!     pub type Foo = u8;
+//! }
+//! ```
+//!
+//! Finally, an alias may be defined over another alias (which is odd but happens). We support those
+//! cases, but require an additional helper attribute on the alias declaration to tell us where
+//! *that* alias is defined. For example:
+//! ```ignore
+//! #[ffi_derive::alias_resolution(crate1_aliases)]
+//! mod aliases_in_crate1 {
+//!     pub type Foo = u8;
+//! }
+//!
+//! #[ffi_derive::alias_resolution(crate2_aliases)]
+//! mod aliases_in_crate2 {
+//!     #[nested_alias="crate1_aliases"]
+//!     pub type Bar = Foo;
+//! }
+//! ```
+//!
+//! ### Remote alias references
+//!
+//! When a type derives an FFI includes a field whose type is an alias defined in a remote crate,
+//! the `ffi_derive` macro invocation just needs to include the helper attribute
+//! `ffi(alias_modules("a_key"))` to tell us the resolution key of the module in which those aliases
+//! are defined. For example:
+//! ```ignore
+//! #[ffi_derive::alias_resolution(crate1_aliases)]
+//! mod aliases_in_crate1 {
+//!     pub type Foo = u8;
+//! }
+//!
+//! #[derive(ffi_derive::FFI), ffi(alias_modules("agrian_types_ids"))]
+//! pub struct SomeTypeInCrate2 {
+//!     pub field: Foo
+//! }
+//! ```
+//!
+//! It's worth noting that there's potential for a couple different issues here. First, if a type
+//! provides multiple keys in `alias_modules`, and an identical alias is defined in each of those
+//! modules, we may interpret the type incorrectly. If that scenario comes up, we can work around it
+//! by moving the helper attribute from the struct to the individual fields (since there we only
+//! need to reference one alias_module at a time), but it gets awfully tedious, so we're not doing
+//! that yet. Second, if a type is renamed when it's imported (as in
+//! `use crate1::aliases::Foo as Meow`), or uses a qualified reference instead of an import (as in
```

Comment:
Maybe `reference` is accurate, but I'm used to seeing `crate::aliases::Foo` as "a fully qualified path", like here: https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#absolute-paths-not-starting-with-crate